### PR TITLE
[WIP] Lazy load panels

### PR DESCRIPTION
**Pull request by ryantxu**
the white whale is back….

This patch offers two approaches for detecting if a panel is currently visible
1. detect scroll and check if top/bottom within 200px of the viewport
2. using IntersectionObserver events

It then sends notice with:
```
panel.events.emit(panel-visibility-changed, true|false);
```

Using the visibility visibility events, this then
A. DashboardPanel (react class) delays putting the panel in the DOM until it is first visible
B. MetricsPanelCtrl delays `onMetricsPanelRefresh` requests until visible

I deployed this in our production environment (not too extreme, but many different clients/browsers).  The behavior for lazy loading panels (A) was spotty... and I removed it.  So far no-one has noticed anything funny with B and the query rate reduced by ~80% :)

### CODE REVIEW

**Comment by mtanda** [comment]
I think Intersection Observer is good API for lazy loading.
https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API

Did you compare scroll event listener and Intersection Observer?

**Comment by ryantxu** [comment]
@mtanda -- the IntersectionObserver API looks good!  I updated the branch so that we can compare the two methods:
```
   this.observer = new PanelObserverScroll();
   if (true) {
      this.observer = new PanelObserverIntersection();
   }
```

Both seem to work -- however I notice lots more lag/slow updates with IntersectionObserver.

**Comment by mtanda** [comment]
@ryantxu Thanks for updating! laggy is not good... I'll check your branch.

**Comment by mtanda** [comment]
I tried to understand how IntersectionObserver work.
https://gist.github.com/mtanda/99b09993b8f5d27cca6b187739a7eb75
I can't understand it well, but I guess it aggregate intersection event, the aggregation cause lag.

One good point of IntersectionObserver is, it doesn't call callback so many times.

File: public/app/features/dashboard/dashgrid/PanelObserver.tsx [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+  }
+
+  //---------------------------------------------------------
+  // Internal
+  //---------------------------------------------------------
+
+  private callback(entries: IntersectionObserverEntry[]) {
+    entries.forEach(entry => {
+      const panel = entry.target['data-garfana-panel'];
+      if (panel.visible !== entry.isIntersecting) {
```
Comment (mtanda) [inline public/app/features/dashboard/dashgrid/PanelObserver.tsx]: If scrolling speed is fast, panel becomes !intersecting => intersecting => !intersecting in one callback.
So, it is better to check all entries before emit event, and emit event if final visible status is changed.

Comment (ryantxu) [reply public/app/features/dashboard/dashgrid/PanelObserver.tsx]: Thanks for the review @mtanda!

I split this into another pull request that just does the visibility changes, but punts lazy loading...

Here is the change you recommended:
https://github.com/grafana/grafana/pull/12531/files#diff-5d8cd443562284b52967432ac7b6b0f2R183

Comment (ryantxu) [reply public/app/features/dashboard/dashgrid/PanelObserver.tsx]: What was your general experience with `IntersectionObserver` vs `scroll` listener?

**Comment by ryantxu** [comment]
@mtanda 
> One good point of IntersectionObserver is, it doesn't call callback so many times.

For the `scroll` option, I added something that only checks every 10px (with a 200px buffer) to avoid LOTS of calculations;
https://github.com/grafana/grafana/pull/12531/files#diff-5d8cd443562284b52967432ac7b6b0f2R122

**Pull request closed without merging**